New LazyControllerFactory

weierophinney opened the pull request:
Inspired by http://circlical.com/blog/2016/3/9/preparing-for-zend-f, this abstract factory/factory can be used to create controller instances for controllers defining constructor dependencies, using the following rules:
- A parameter named `$config` typehinted as an array will receive the application "config" service (i.e., the merged configuration).
- Parameters type-hinted against array, but not named `$config` will be injected with an empty array.
- Scalar parameters will be resolved as null values.
- If a service cannot be found for a given typehint, the factory will raise an exception detailing this.
- Some services provided by Zend Framework components do not have entries based on their class name (for historical reasons); the factory contains a map of these class/interface names to the corresponding service name to allow them to resolve.

`$options` passed to the factory are ignored in all cases, as we cannot make assumptions about which argument(s) they might replace.

As it implements zend-servicemanager's v3 AbstractFactoryInterface, it may be used as either an abstract factory, or by mapping controller class names to the factory.

weierophinney committed:
New LazyControllerFactory

Inspired by http://circlical.com/blog/2016/3/9/preparing-for-zend-f,
this abstract factory/factory can be used to create controller
instances for controllers defining constructor dependencies, using
the following rules:

- A parameter named `$config` typehinted as an array will receive the
  application "config" service (i.e., the merged configuration).
- Parameters type-hinted against array, but not named `$config` will
  be injected with an empty array.
- Scalar parameters will be resolved as null values.
- If a service cannot be found for a given typehint, the factory will
  raise an exception detailing this.
- Some services provided by Zend Framework components do not have
  entries based on their class name (for historical reasons); the
  factory contains a map of these class/interface names to the
  corresponding service name to allow them to resolve.

`$options` passed to the factory are ignored in all cases, as we cannot
make assumptions about which argument(s) they might replace.

As it implements zend-servicemanager's v3 AbstractFactoryInterface, it
may be used as either an abstract factory, or by mapping controller
class names to the factory.

Ocramius commented on src/Controller/LazyControllerFactory.php:
```diff
@@ -0,0 +1,154 @@
+     * @var string[]
+     */
+    private $aliases = [
+        'Zend\Console\Adapter\AdapterInterface'     => 'ConsoleAdapter',
```
`::class` all these, plz

snapshotpl replied:
`::class` need to install all repositories, because need this classes

Ocramius replied:
No, `::class` does not cause autoloading.

weierophinney replied:
@snapshotpl Not true; PHP will resolve the class name using `::class` even if it cannot autoload the class.

Ocramius replied:
You can use `::class` even with non-existing classes

snapshotpl replied:
Interesting. So `Zend\Console\Adapter\AdapterInterface::class` will be `Zend\Console\Adapter\AdapterInterface` string even I autoload it or not?

Ocramius commented on src/Controller/LazyControllerFactory.php:
```diff
@@ -0,0 +1,154 @@
+ * `$options` passed to the factory are ignored in all cases, as we cannot
+ * make assumptions about which argument(s) they might replace.
+ */
+class LazyControllerFactory implements AbstractFactoryInterface
```
s/Factory/AbstractFactory

Ocramius commented on src/Controller/LazyControllerFactory.php:
```diff
@@ -0,0 +1,154 @@
+            return false;
+        }
+
+        $implements = class_implements($requestedName);
```
No need for this assignment

Ocramius commented on src/Controller/LazyControllerFactory.php:
```diff
@@ -0,0 +1,154 @@
+     */
+    public function canCreate(ContainerInterface $container, $requestedName)
+    {
+        if (! is_string($requestedName) || ! class_exists($requestedName)) {
```
can `$requestedName` ever be a non-string?

weierophinney replied:
I've seen `null` a few times, though primarily when used in a servicemanager v2 context. I'll remove that check.

weierophinney replied:
To clarify: zend-mvc v3 _requires_ zend-servicemanager v3, and the class is implementing only the v3 interface. Thus, the check is superfluous.

weierophinney committed:
Added documentation for LazyControllerFactory

snapshotpl commented on src/Controller/LazyControllerFactory.php:
```diff
@@ -0,0 +1,154 @@
+     *
+     * @var string[]
+     */
+    private $aliases = [
```
It will be nice to provide custom aliases

weierophinney replied:
I'll make it protected so this factory can be extended, then.

weierophinney committed:
Use `::class` for aliased classes in LazyControllerFactory

- Per @Ocramius

snapshotpl commented on src/Controller/LazyControllerFactory.php:
```diff
@@ -0,0 +1,154 @@
+            $parameters[] = $container->get($type);
+        }
+
+        return $reflectionClass->newInstanceArgs($parameters);
```
Why not use http://php.net/manual/en/migration56.new-features.php#migration56.new-features.splat if we support php `^5.6` ?

weierophinney committed:
Renamed factory to incorporate verbiage "Abstract"

- per @Ocramius

weierophinney committed:
Remove test for non-string $requestedName

cannot happen with version 3 servicemanager.

snapshotpl commented on src/Controller/LazyControllerFactory.php:
```diff
@@ -0,0 +1,164 @@
+            }
+
+            $type = $parameter->getClass()->getName();
+            $type = array_key_exists($type, $this->aliases) ? $this->aliases[$type] : $type;
```
`isset` is faster, right?

weierophinney committed:
Made $aliases protected (not private)

- Allows extending in order to add more aliases.

snapshotpl commented on src/Controller/LazyControllerAbstractFactory.php:
```diff
@@ -0,0 +1,165 @@
+     * required for those services with no entry based on the class/interface
+     * name.
+     *
+     * Extend the class if you wish to add to the list.
```
Can't we get this list from configuration?

weierophinney replied:
I'd rather not add a configuration point for it. This currently lists all "special case" service names, and we recommend that users utilize fully qualified interface and/or class names for their service names.

If a user really wants to add more, they should have to think twice before doing so.

snapshotpl replied:
No I see that https://github.com/zendframework/zend-mvc/pull/165/files#diff-898a3a94db9fcef93eeeb26a23397ecaR139 will resolve many cases. Anyway this feature it's for really lazy developers, and never will be cover all cases

weierophinney replied:
Precisely. If you need something more robust, or with more features, you should likely create a custom factory at that point, not extend this one. :smile:

weierophinney committed:
Use splat operator

- Updated instantiation to use splat operator instead of reflection
- switch from array_key_exists to isset

Both per @snapshotpl

weierophinney commented:
@snapshotpl Thanks for reminding me about the splat operator; still not used to being able to use PHP 5.6 features!

snapshotpl commented:
That's why we goes into 5.6!, right?

weierophinney committed:
Refactor: extract method

Extracted parameter resolution to a new method, and modified factory to
use array_map with that method to create the list of parameters for
instantiating the controller.

snapshotpl commented on src/Controller/LazyControllerAbstractFactory.php:
```diff
@@ -0,0 +1,177 @@
+            return new $requestedName();
+        }
+
+        $parameters = array_map(function (ReflectionParameter $parameter) use ($container, $requestedName) {
```
I read about performance `array_map` and `foreach` and looks a little bit slower. `use` also decrease readability for me. Maybe `foreach` will be better?

weierophinney replied:
Micro-optimizations with the size of the arrays we have here. But I agree on the readability part; really wish you could do callables using private methods if invoked within the context of the class.

weierophinney replied:
I've refactored this now. `resolveParameter()` now returns a closure over the container and requested name, leaving:

``` php
$parameters = array_map(
    $this->resolveParameter($container, $requestedName),
    $reflectionParameters
);
```

weierophinney committed:
Refactor array_map

`resolveParameter()` now accepts the container and requested name, and
returns a callback for resolving a parameter to a value, closing over
the two provided values. This simplifies the array_map call, making it
more readable.

Ocramius commented:
@weierophinney I just realized that this PR has nothing to do with laziness, but is instead centered on "automatic di". Should I open an issue about it, or can you still work on it?

weierophinney commented:
@ocramius — What's your idea? A rename? Or a more generic offering in
zend--servicemanager?
On Jun 23, 2016 7:39 PM, "Marco Pivetta" notifications@github.com wrote:

> @weierophinney https://github.com/weierophinney I just realized that
> this PR has nothing to do with laziness, but is instead centered on
> "automatic di". Should I open an issue about it, or can you still work on
> it?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/165#issuecomment-228226437,
> or mute the thread
> https://github.com/notifications/unsubscribe/AABlV0WWRFtIntt4GKtWu80bvyqAgGISks5qOyddgaJpZM4I9CNX
> .

Ocramius commented:
A rename is sufficient. Just want to avoid messy situations with
documentation later on
On Jun 24, 2016 03:49, "weierophinney" notifications@github.com wrote:

> @ocramius — What's your idea? A rename? Or a more generic offering in
> zend--servicemanager?
> On Jun 23, 2016 7:39 PM, "Marco Pivetta" notifications@github.com wrote:
> 
> > @weierophinney https://github.com/weierophinney I just realized that
> > this PR has nothing to do with laziness, but is instead centered on
> > "automatic di". Should I open an issue about it, or can you still work on
> > it?
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> > https://github.com/zendframework/zend-mvc/pull/165#issuecomment-228226437
> > ,
> > or mute the thread
> > <
> > https://github.com/notifications/unsubscribe/AABlV0WWRFtIntt4GKtWu80bvyqAgGISks5qOyddgaJpZM4I9CNX
> > 
> > .
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/165#issuecomment-228235980,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAJakDtk9DVdI5rx6WcKvAaW1Ch_8r7Iks5qOzengaJpZM4I9CNX
> .
